clang/lib/DPCT/BarrierFenceSpaceAnalyzer.cpp

```diff
@@ -7,311 +7,570 @@
+      FoundBO = true;
+    } else if (!FoundDeref && UOInBO &&
+               (UOInBO->getOpcode() == UnaryOperatorKind::UO_Deref)) {
+      FoundDeref = true;
+    } else if (!FoundDeref && ASEInBO &&
+               (ASEInBO->getBase() == Current.get<Expr>())) {
+      FoundDeref = true;
+    }
+
+    Current = Parents[0];
+    Parents = Context.getParents(Current);
   }
 
-  // Special case for atomicAdd.
-  // If the Pointer is the first arg of atomicAdd function, return true.
-  if (auto CE = dyn_cast<clang::CallExpr>(P)) {
-    if (auto FD = CE->getDirectCallee()) {
-      std::string FuncName = FD->getNameInfo().getName().getAsString();
-      if (FuncName == "atomicAdd" && (CE->getArg(0) == Pointer) &&
-          !isUserDefinedDecl(FD)) {
-        return true;
-      }
+  if (PVD) {
+    if (PVD->getType()->isPointerType()) {
+      if (PVD->getType()->getPointeeType().isConstQualified())
+        return AccessMode::Read;
+      return AccessMode::ReadWrite;
+    }
+    if (PVD->getType()->isLValueReferenceType()) {
+      if (PVD->getType().getNonReferenceType().isConstQualified())
+        return AccessMode::Read;
+      return AccessMode::ReadWrite;
     }
+    return AccessMode::Read;
   }
-  return false;
+
+  if (BO) {
+    if (BO->getOpcode() == BinaryOperatorKind::BO_Assign) {
+      return AccessMode::Write;
+    }
+    return AccessMode::ReadWrite;
+  }
+
+  // No BO or Call in ancestor
+  return AccessMode::Read;
 }
 
-bool clang::dpct::BarrierFenceSpaceAnalyzer::canSetLocalFenceSpace(
-    const clang::CallExpr *CE) {
-  if (CE->getBeginLoc().isMacroID() || CE->getEndLoc().isMacroID())
-    return false;
-  auto FD = dpct::DpctGlobalInfo::findAncestor<clang::FunctionDecl>(CE);
-  if (!FD)
-    return false;
-  if (!FD->hasAttr<clang::CUDAGlobalAttr>())
-    return false;
-  if (FD->getTemplateSpecializationKind() !=
-          TemplateSpecializationKind::TSK_Undeclared ||
-      FD->getDescribedFunctionTemplate()) {
+bool clang::dpct::BarrierFenceSpaceAnalyzer::isPotentialGlobalMemoryAccess(
+    std::shared_ptr<DeviceFunctionInfo> DFI, bool IsInGlobalFunction) {
+  if (!DFI)
     return false;
+
+  const MemVarInfoMap &MVIM =
+      DFI->getVarMap().getMap(MemVarInfo::VarScope::Global);
+  for (const auto &VarInfo : MVIM) {
+    auto Attr = VarInfo.second->getAttr();
+    if (Attr == MemVarInfo::VarAttrKind::Device ||
+        Attr == MemVarInfo::VarAttrKind::Managed)
+      return true;
+  }
+  for (const auto &Call : DFI->getCallExprMap()) {
+    if (Call.second->getName() == "__syncthreads" && !IsInGlobalFunction) {
+      // Currently we do not support the analysis for __syncthreads in
+      // __device__ function
+      return false;
+    }
+    if (isPotentialGlobalMemoryAccess(Call.second->getFuncInfo(), false))
+      return true;
+  }
+  return false;
+}
+
+clang::dpct::BarrierFenceSpaceAnalyzerResult
+clang::dpct::BarrierFenceSpaceAnalyzer::canSetLocalFenceSpace(
```

Comment:
```suggestion
clang::dpct::BarrierFenceSpaceAnalyzer::analyze(
```

Reply:
Function L342 is too big,  difficult to list the whole logic there,  please refactor it by introducing more sub functions.

Reply:
done